Update to MV3: DeclarativeContent API (#2261)
GoogleChrome/developer.chrome.com: site/en/docs/extensions/reference/declarativeContent/index.md

api: declarativeContent
---

{% Aside %}

To transition from pageaction to action, see [Emulating pageActions with declarativeContent][emulating-page-actions]

{% endAside %}

## Usage

The Declarative Content API allows you to enable your extension's [action][api-action] depending on

> the URL of a web page and the CSS selectors its content matches, without needing to add [host

Reviewer: This phrasing is a bit odd. Does it mean that a URL triggers an action?

> permission][docs-host-perm] or inject a [content script][docs-content-scripts]. 

Author: Yes! Good catch :)

> Use the [activeTab][docs-activetab] permission to interact with a page after the user clicks on the

Reviewer: ```suggestion
Use the [activeTab][docs-activetab] permission to interact with a page after the user clicks the
```

extension's action (toolbar icon).

## Rules

> As a [declarative API][api-declarative], this API lets you register rules on the

Reviewer: ```suggestion
The [declarative API][api-declarative] lets you register rules on the
```

> [`onPageChanged`][event-onpagechanged] event object which take an action

Reviewer: >lets you register rules on the on the `onPageChanged` event object

This phrase feels like its cramming too much in formation in too small a space and I can't disentangle it. I would expect something called `onPageChanged` to be defined on a target object, take a custom method (a callback, though it's not usually called that in JS) and for the method to get passed an event object. This description reads like it's conflated the target, the callback, and the event into one thing. It's confusing.

Author: Agreed! I restructured this section to detangle key concepts. Let me know what you think.

api: declarativeContent
---

{% Aside %}

To transition from page action to action, see [Emulating pageActions with
declarativeContent][emulating-page-actions]

{% endAside %}

## Usage

{% Aside 'key-term' %}

> An [_extension's action_][api-action] controls the extension's toolbar icon.

Reviewer: ```suggestion
An [extension's _action_][api-action] controls the extension's toolbar icon.
```

{% endAside %}

The Declarative Content API allows you to enable your extension's action depending on the URL of a

> web page, or if a CSS selector matches an element on the page, without needing to 

Author: @jpmedley How does this sound?

Reviewer: Sounds good.

> web page and the CSS selectors, without needing to add [host permissions][docs-host-perm] or inject a

Reviewer: >and the CSS selectors.

What is this telling me? I don't understand this part.

Author: I just made a change, let me know what you think

[content script][docs-content-scripts]. 

Use the [activeTab][docs-activetab] permission to interact with a page after the user clicks on the
extension's action.

## Rules

Rules consists of conditions and actions. If any of the conditions is fulfilled, all actions are
executed. The actions are [`setIcon`][type-set-icon] and [`showAction`][type-show-action].

The [`PageStateMatcher`][type-page-state-matcher] matches web pages if and only if all listed
criteria are met. It can match a [page url][section-pageurl], a [css compound selector][section-css]
or the [bookmarked state][section-bookmarked] of a page. The following rule enables
the extension's action on Google pages when a password field is present:

```js
let rule1 = {
  conditions: [
    new chrome.declarativeContent.PageStateMatcher({
      pageUrl: { hostSuffix: '.google.com', schemes: ['https'] },
      css: ["input[type='password']"]
    })
  ],
  actions: [ new chrome.declarativeContent.ShowAction() ]
};

```

{% Aside 'success' %}

All conditions and actions are created via a constructor as shown in the example above.

{% endAside %}

> In order to also enable the extension's action for Google sites with a video, you can add a second

Reviewer: ```suggestion
To also enable the extension's action for Google sites with a video, you can add a second
```

condition, as each condition is sufficient to trigger all specified actions:

```js
let rule2 = {
  conditions: [
    new chrome.declarativeContent.PageStateMatcher({
      pageUrl: { hostSuffix: '.google.com', schemes: ['https'] },
      css: ["input[type='password']"]
    }),
    new chrome.declarativeContent.PageStateMatcher({
      css: ["video"]
    })
  ],
  actions: [ new chrome.declarativeContent.ShowAction() ]
};
```

The [`onPageChanged`][event-onpagechanged] event tests whether any rule has at least one fulfilled

> condition and executes the actions. Rules are persistent across browsing sessions; therefore, during

Reviewer: ```suggestion
condition and executes the actions. Rules persist across browsing sessions; therefore, during
```

{% Aside %}

You should always register or unregister rules in bulk because each of these operations recreates
internal data structures. This re-creation is computationally expensive but facilitates a faster
matching algorithm.

{% endAside %}

With the [activeTab][docs-activetab] permission, your extension will not display any permission
warnings and when the user clicks the extension action, it will only run on relevant pages.

## Page URL Matching {: #page-url}

> The [`PageStateMatcher.pageurl`][type-matcher-url] matches if the URL criterias are fulfilled. The

Reviewer: ```suggestion
The [`PageStateMatcher.pageurl`][type-matcher-url] matches when the URL criteria are fulfilled. The
```

> most common criterias are a concatination of either host, path, or url, followed by Contains, Equals, Prefix, or

Reviewer: ```suggestion
most common criteria are a concatenation of either host, path, or url, followed by Contains, Equals, Prefix, or
```

Suffix. The following table contains a few examples:

| Criteria                                  | Matches                         |
|-------------------------------------------|---------------------------------|
| `{ hostSuffix: 'google.com' }`            | All Google URLs                 |
| `{ pathPrefix: '/docs/extensions'` }      | Extension docs URLs             |
| `{ urlContains: 'developer.chrome.com'` } | All chrome developers docs URLs |

> All criteria are case sensitive. For complete list of criterias, see [UrlFilter][type-urlfilter].

Reviewer: ```suggestion
All criteria are case sensitive. For a complete list of criteria, see [UrlFilter][type-urlfilter].
```

## CSS Matching {: #css}

> [`PageStateMatcher.css`][type-matcher-css] conditions must be _[comp``ound selectors][w3-compound]_,

Reviewer: ```suggestion
[`PageStateMatcher.css`][type-matcher-css] conditions must be _[compound selectors][w3-compound]_,
```
